portality/view/oaipmh.py

```diff
@@ -145,7 +145,15 @@ def decode_set_spec(setspec):
     return decoded
 
 
-def make_resumption_token(metadata_prefix=None, from_date=None, until_date=None, oai_set=None, start_number=None):
+def get_start_after(docs):
+    last_date = docs[-1].get("last_updated")
+    count = 0
+    for doc in docs:
+        if doc.get("last_updated") == last_date:
+            count += 1
+    return (last_date, count)
```

Comment:
I was thinking a bit more about this, and wondering if we might encounter any issues, and one possibility came to mind.

Sometimes we run migrations on the database which touch all of the items.  As a result, their last_updated date gets refreshed, and for a time the entire index will have last_updated dates all within a few seconds of each other. Could this be a problem?  It might cause the same deep paging effect, until the last_updated dates get more spread out by ongoing actions.

We don't do this very often, and given how long migrations can take to run it's unlikely that the entire index would have the same last_updated date, so the problem could never be as bad as it is now, but is there anything else clever we can do to order the items or filter the items to avoid this date-searching issue.  The newer ES version's "search_after" /would/ get us around this issue, so a possible +1 for a back-end upgrade there.

Reply:
Ah, in fact, I'll go you one further - I can see a bug that this might cause.  If you get to two pages with all the same last_updated dates, doesn't this make it impossible to get past the second page?

So, for example, suppose there are 200 records with the same last_updated date, and we have page sizes of 10.  On the first page there will be 10 results, all with date A.  The start_after value will then be (A, 11).  So the next query will get records 11 - 20.  These will also have date A, so the start_after value will be (A, 11) again, as the date is the same and there are 10 results all with that date.

I have some thoughts on a fix, though not well worked through yet:

* We can get the correct value for `start_after[1]` if we can keep track of how many previous records (including on previous pages) have the same value for last_updated
* We can probably encode this information in the resumption token
* We'll need to check the final record on the current page, and if its last_updated date is the same as the one that we started with, then sum the count of records on the page with the sum of all previous records with that last_updated date, and send that value forwards.  If it is not the same, then reset that cross-page counter back to 0, and then start again from the current page's offset.

Reply:
On the other point, another thing we could do is increase the last_updated granularity to the millisecond scale, which would minimise the number of records which appeared to have the same date.  Right now it is just stored to the second.

Reply:
Upgrading ES to use search_after will not help in case of migrations, as it uses the sort functionality to figure out the next record. Increasing the granularity of time to milliseconds should help

Ah yes, If the new record_offset equals the number of records, I also need to add the record_offset from the existing search. Thanks for spotting that. I will make that change.

Reply:
This function looks good to me now